/*
    Copyright Â© 2023 Fernando Porrino Serrano
    Third party software licenses can be found at /docs/credits/credits.md

    This file is part of AutoCheck.

    AutoCheck is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AutoCheck is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with AutoCheck.  If not, see <https://www.gnu.org/licenses/>.
*/

using System.IO;
using System.Collections.Generic;

using NUnit.Framework;

namespace AutoCheck.Test.Real
{    
    [Parallelizable(ParallelScope.All)]    
    public class Views : Test
    { 
        public Views(): base("script/real"){
        }        

        [NonParallelizable] //cannot interfear with the batch one, shares users and roles
        [Test, Category("FullScriptViews"), Category("Real"), Category("Local")]
        public void Real_VIEWS_SCRIPT_SINGLE_1()
        {
            //TODO: the SQL script should create also users and restore the privileges.

            //Note: cannot be splitted or parallelized, must be done sequentualy because users and roles are shared alog the postgres instance.
            var conn = new AutoCheck.Core.Connectors.Postgres("localhost", "empresa", "postgres", "postgres");
            if(!conn.ExistsDataBase()) conn.CreateDataBase();

            //Restoring users
            if(!conn.ExistsUser("it")) conn.CreateUser("it", "it");
            else conn.DropUserMembership("it");           
            
            //Restoring database
            var privatePath = Path.Combine(SamplesRootFolder, "private", "views");
            conn.CreateDataBase(Path.Combine(privatePath, "restore.sql"), true);

            //Original status
            var s = new AutoCheck.Core.Script(GetScriptFile(Core.Script.ExecutionMode.SINGLE, "views.yaml"));
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...ERROR:\n         -Expected -> True; Found -> False\r\n      Checking view definition...ERROR:\n         -Value cannot be null. (Parameter 'compared')\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...ERROR:\n         -Unable to run te given SQL query 'INSERT INTO gerencia.responsables (nom_fabrica, nom_responsable, cognoms_responsable) VALUES ('NEW FACTORY NAME 4', 'NEW EMPLOYEE NAME 10', 'NEW EMPLOYEE SURNAME 10')'. Please, check the inner exception for further details. \r\n         ---> 42P01: relation \"gerencia.responsables\" does not exist \n          \n         POSITION: 13\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=10 AND nom_responsable='NEW EMPLOYEE NAME 10' AND cognoms_responsable='NEW EMPLOYEE SURNAME 10''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for new content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='NEW FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for new content in produccio.fabrique...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 10', cognoms_responsable='UPDATED EMPLOYEE SURNAME 10' WHERE id_responsable=10'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=10 AND nom_responsable='UPDATED EMPLOYEE NAME 10' AND cognoms_responsable='UPDATED EMPLOYEE SURNAME 10''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 4' WHERE id_fabrica=4'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='UPDATED FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=11'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 0 / 10", s.Output.ToString());

            //Solve question 1
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex1.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...ERROR:\n         -Unable to run te given SQL query 'INSERT INTO gerencia.responsables (nom_fabrica, nom_responsable, cognoms_responsable) VALUES ('NEW FACTORY NAME 6', 'NEW EMPLOYEE NAME 12', 'NEW EMPLOYEE SURNAME 12')'. Please, check the inner exception for further details. \r\n         ---> 55000: cannot insert into view \"responsables\" \n          \n         DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for new content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for new content in produccio.fabrique...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 12', cognoms_responsable='UPDATED EMPLOYEE SURNAME 12' WHERE id_responsable=12'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot update view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 6' WHERE id_fabrica=6'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot update view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=13'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=7'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 2 / 10", s.Output.ToString());

            //Solve question 2
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex2.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 15', cognoms_responsable='UPDATED EMPLOYEE SURNAME 15' WHERE id_responsable=15'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot update view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 9' WHERE id_fabrica=9'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot update view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 1; Found -> 0\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=16'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=10'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 4 / 10", s.Output.ToString());

            //Solve question 3
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex3.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in rrhh.empleats...OK\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=19'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=13'. Please, check the inner exception for further details. \r\n            ---> 55000: cannot delete from view \"responsables\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 6 / 10", s.Output.ToString());

            //Trying to solve the exercise removing the employee data
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex4_delete_v1.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in rrhh.empleats...OK\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=22'. Please, check the inner exception for further details. \r\n            ---> 23503: update or delete on table \"empleats\" violates foreign key constraint \"fabriques_id_responsable_fkey\" on table \"fabriques\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=16'. Please, check the inner exception for further details. \r\n            ---> 23503: update or delete on table \"empleats\" violates foreign key constraint \"fabriques_id_responsable_fkey\" on table \"fabriques\" \n             \n            DETAIL: Detail redacted as it may contain sensitive data. Specify 'Include Error Detail' in the connection string to include this information.\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 6 / 10", s.Output.ToString());

            //Trying to solve the exercise removing the factory data
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex4_delete_v2.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in rrhh.empleats...OK\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in from gerencia.responsables...OK\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...OK\r\n         Checking for non-deleted content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 6 / 10", s.Output.ToString());
                        
            //Solve question 4
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex4.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in rrhh.empleats...OK\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in from gerencia.responsables...OK\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...OK\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 9 / 10", s.Output.ToString());

            //Solve question 5
            conn.ImportSqlDump(Path.Combine(privatePath, "solve_ex5.sql"));
            s.Start();
            Assert.AreEqual("Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\nRunning on single mode:\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...OK\r\n      Checking view definition...OK\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...OK\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in rrhh.empleats...OK\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...OK\r\n         Checking for new content in produccio.fabrique...OK\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in rrhh.empleats...OK\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in from gerencia.responsables...OK\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...OK\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...OK\r\n         Checking for updated content in gerencia.responsables...OK\r\n         Checking for updated content in produccio.fabriques...OK\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...OK\r\n      Looking for permissions into the schema gerencia...OK\r\n\r\n   TOTAL SCORE: 10 / 10", s.Output.ToString());
        }       

        [NonParallelizable] //cannot interfear with the single one, shares users and roles
        [Test, Category("FullScriptViews"), Category("Real"), Category("Local")]
        public void Real_VIEWS_SCRIPT_BATCH_1()
        {             
            var s = new AutoCheck.Core.Script(GetScriptFile(Core.Script.ExecutionMode.BATCH, "views.yaml"), false);      
            s.OverrideTarget(new Dictionary<string, string>(){ {"path", "{$APP_FOLDER_PATH}/../test/samples/private/views"} }); 
            s.Start();

            var expected = "Running script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\n   Extracting files at User_Folder_1:\r\n      Extracting the file var.zip... OK\r\n\r\n   Extracting files at User_Folder_2:\r\n      Extracting the file NULL... OK\r\n\r\n   Extracting files at User_Folder_3:\r\n      Extracting the file NULL... OK\r\n\r\n   Starting the copy detector for SqlLog:\r\n      Looking for potential copies within User_Folder_1... OK\r\n      Looking for potential copies within User_Folder_2... OK\r\n      Looking for potential copies within User_Folder_3... OK\r\n\r\nRunning on batch mode for User_Folder_1:\r\n\r\n   Restoring BBDD permissions:\r\n      Dropping user it... OK\r\n      Creating user it... OK\r\n\r\n   Setting up SQL dump:\r\n      Getting student name... OK\r\n      Getting database name... OK\r\n      Looking for the SQL dump... OK\r\n      Replacing the database name... OK\r\n      Replacing the database connection... OK\r\n      Storing changes... OK\r\n\r\n   Importing SQL dump:\r\n      Restoring database empresa-permissions-User    using the file empresa_Alejandro_Mallen.sql... OK\r\n\r\n   No potential copy detected for User_Folder_1/postgresql-12-main.log:\r\n      Match score with User_Folder_2/postgresql-12-main.log... 47,41 %\r\n      Match score with User_Folder_3/postgresql-12-main.log... 42,90 %\r\n\r\n   No potential copy has been detected.\r\n\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...ERROR:\n         -Expected -> True; Found -> False\r\n      Checking view definition...ERROR:\n         -Value cannot be null. (Parameter 'compared')\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...ERROR:\n         -Unable to run te given SQL query 'INSERT INTO gerencia.responsables (nom_fabrica, nom_responsable, cognoms_responsable) VALUES ('NEW FACTORY NAME 4', 'NEW EMPLOYEE NAME 11', 'NEW EMPLOYEE SURNAME 11')'. Please, check the inner exception for further details. \r\n         ---> 42P01: relation \"gerencia.responsables\" does not exist \n          \n         POSITION: 13\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=11 AND nom_responsable='NEW EMPLOYEE NAME 11' AND cognoms_responsable='NEW EMPLOYEE SURNAME 11''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for new content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='NEW FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for new content in produccio.fabrique...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 11', cognoms_responsable='UPDATED EMPLOYEE SURNAME 11' WHERE id_responsable=11'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=11 AND nom_responsable='UPDATED EMPLOYEE NAME 11' AND cognoms_responsable='UPDATED EMPLOYEE SURNAME 11''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 4' WHERE id_fabrica=4'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='UPDATED FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=12'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 0 / 10\r\n\r\nRunning script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\n   Extracting files at User_Folder_1:\r\n      Extracting the file var.zip... OK\r\n\r\n   Extracting files at User_Folder_2:\r\n      Extracting the file NULL... OK\r\n\r\n   Extracting files at User_Folder_3:\r\n      Extracting the file NULL... OK\r\n\r\n   Starting the copy detector for SqlLog:\r\n      Looking for potential copies within User_Folder_1... OK\r\n      Looking for potential copies within User_Folder_2... OK\r\n      Looking for potential copies within User_Folder_3... OK\r\n\r\nRunning on batch mode for User_Folder_2:\r\n\r\n   Restoring BBDD permissions:\r\n      Dropping user it... OK\r\n      Creating user it... OK\r\n\r\n   Setting up SQL dump:\r\n      Getting student name... OK\r\n      Getting database name... OK\r\n      Looking for the SQL dump... OK\r\n      Replacing the database name... OK\r\n      Replacing the database connection... OK\r\n      Storing changes... OK\r\n\r\n   Importing SQL dump:\r\n      Restoring database empresa-permissions-User    using the file empresa_pratik_Examen_Patel.sql... OK\r\n\r\n   No potential copy detected for User_Folder_2/postgresql-12-main.log:\r\n      Match score with User_Folder_1/postgresql-12-main.log... 47,41 %\r\n      Match score with User_Folder_3/postgresql-12-main.log... 42,23 %\r\n\r\n   No potential copy has been detected.\r\n\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...ERROR:\n         -Expected -> True; Found -> False\r\n      Checking view definition...ERROR:\n         -Value cannot be null. (Parameter 'compared')\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...ERROR:\n         -Unable to run te given SQL query 'INSERT INTO gerencia.responsables (nom_fabrica, nom_responsable, cognoms_responsable) VALUES ('NEW FACTORY NAME 4', 'NEW EMPLOYEE NAME 11', 'NEW EMPLOYEE SURNAME 11')'. Please, check the inner exception for further details. \r\n         ---> 42P01: relation \"gerencia.responsables\" does not exist \n          \n         POSITION: 13\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=11 AND nom_responsable='NEW EMPLOYEE NAME 11' AND cognoms_responsable='NEW EMPLOYEE SURNAME 11''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for new content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='NEW FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for new content in produccio.fabrique...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 11', cognoms_responsable='UPDATED EMPLOYEE SURNAME 11' WHERE id_responsable=11'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=11 AND nom_responsable='UPDATED EMPLOYEE NAME 11' AND cognoms_responsable='UPDATED EMPLOYEE SURNAME 11''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 4' WHERE id_fabrica=4'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='UPDATED FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=12'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 0 / 10\r\n\r\nRunning script ASIX - M02 (UF3): Views Assignment (v1.1.1.0):\r\n   Extracting files at User_Folder_1:\r\n      Extracting the file var.zip... OK\r\n\r\n   Extracting files at User_Folder_2:\r\n      Extracting the file NULL... OK\r\n\r\n   Extracting files at User_Folder_3:\r\n      Extracting the file NULL... OK\r\n\r\n   Starting the copy detector for SqlLog:\r\n      Looking for potential copies within User_Folder_1... OK\r\n      Looking for potential copies within User_Folder_2... OK\r\n      Looking for potential copies within User_Folder_3... OK\r\n\r\nRunning on batch mode for User_Folder_3:\r\n\r\n   Restoring BBDD permissions:\r\n      Dropping user it... OK\r\n      Creating user it... OK\r\n\r\n   Setting up SQL dump:\r\n      Getting student name... OK\r\n      Getting database name... OK\r\n      Looking for the SQL dump... OK\r\n      Replacing the database name... OK\r\n      Replacing the database connection... OK\r\n      Storing changes... OK\r\n\r\n   Importing SQL dump:\r\n      Restoring database empresa-permissions-User    using the file empresa_Aure_Gonzales.sql... OK\r\n\r\n   No potential copy detected for User_Folder_3/postgresql-12-main.log:\r\n      Match score with User_Folder_1/postgresql-12-main.log... 42,90 %\r\n      Match score with User_Folder_2/postgresql-12-main.log... 42,23 %\r\n\r\n   No potential copy has been detected.\r\n\r\n   Question 1 [2 points] - View creation:\r\n      Checking view creation...ERROR:\n         -Expected -> True; Found -> False\r\n      Checking view definition...ERROR:\n         -Value cannot be null. (Parameter 'compared')\r\n\r\n   Question 2 [2 points] - Insert rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into gerencia.responsables...ERROR:\n         -Unable to run te given SQL query 'INSERT INTO gerencia.responsables (nom_fabrica, nom_responsable, cognoms_responsable) VALUES ('NEW FACTORY NAME 4', 'NEW EMPLOYEE NAME 10', 'NEW EMPLOYEE SURNAME 10')'. Please, check the inner exception for further details. \r\n         ---> 42P01: relation \"gerencia.responsables\" does not exist \n          \n         POSITION: 13\r\n\r\n      Question 2.1 [1 point] - Insert rule (empleats):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=10 AND nom_responsable='NEW EMPLOYEE NAME 10' AND cognoms_responsable='NEW EMPLOYEE SURNAME 10''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for new content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 2.2 [1 point] - Insert rule (fabriques):\r\n         Checking for new content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='NEW FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for new content in produccio.fabrique...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 3 [2 points] - Update rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 3.1 [1 point] - Update rule (empleats):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_responsable='UPDATED EMPLOYEE NAME 10', cognoms_responsable='UPDATED EMPLOYEE SURNAME 10' WHERE id_responsable=10'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_responsable=10 AND nom_responsable='UPDATED EMPLOYEE NAME 10' AND cognoms_responsable='UPDATED EMPLOYEE SURNAME 10''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for updated content in rrhh.empleats...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n      Question 3.2 [1 point] - Update rule (fabriques):\r\n         Updating data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'UPDATE gerencia.responsables SET nom_fabrica='UPDATED FACTORY NAME 4' WHERE id_fabrica=4'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 8\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=4 AND nom_fabrica='UPDATED FACTORY NAME 4''. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 4 [3 points] - Delete rule:\r\n      Loading demo data for produccio.fabriques...OK\r\n      Loading demo data for rrhh.empleats...OK\r\n      Inserting demo data into rrhh.empleats...OK\r\n      Inserting demo data into produccio.fabriques...OK\r\n\r\n      Question 4.1 [2 points] - Delete rule (empleats):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_responsable=11'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_responsable) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 35\r\n         Checking for non-deleted content in from rrhh.empleats...OK\r\n         Checking for updated content in from produccio.fabriques...ERROR:\n            -Expected -> 0; Found -> 1\r\n         Checking for non-deleted content in produccio.fabriques...OK\r\n\r\n      Question 4.2 [1 point] - Delete rule (fabriques):\r\n         Deleting data from gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'DELETE FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 13\r\n         Checking for updated content in gerencia.responsables...ERROR:\n            -Unable to run te given SQL query 'SELECT COUNT(id_fabrica) FROM gerencia.responsables WHERE id_fabrica=5'. Please, check the inner exception for further details. \r\n            ---> 42P01: relation \"gerencia.responsables\" does not exist \n             \n            POSITION: 31\r\n         Checking for updated content in produccio.fabriques...ERROR:\n            -Expected -> 1; Found -> 0\r\n\r\n   Question 5 [1 point] - Permissions:\r\n      Looking for permissions into the view gerencia.responsables...ERROR:\n         -Expected -> r; Found -> \r\n      Looking for permissions into the schema gerencia...ERROR:\n         -Expected -> U; Found -> \r\n\r\n   TOTAL SCORE: 0 / 10";
            Assert.AreEqual(expected, s.Output.ToString());
        }
    }
}